packages/patterns/src/keys/keycollection-operators.js

```diff
@@ -0,0 +1,226 @@
+// @ts-check
+import {
+  assertRankSorted,
+  compareAntiRank,
+  makeFullOrderComparatorKit,
+  sortByRank,
+} from '@endo/marshal';
+import { makeIterator, makeArrayIterator } from '../utils.js';
+
+/** @typedef {import('@endo/marshal').RankCompare} RankCompare */
+/** @typedef {import('../types').KeyComparison} KeyComparison */
+/** @typedef {import('../types').KeyCompare} KeyCompare */
+/** @typedef {import('../types').FullCompare} FullCompare */
+/** @typedef {import('../types').KeyCollection} KeyCollection */
+
+const { quote: q, Fail } = assert;
+
+/**
+ * Refines a sequence of entries that is already sorted over its keys by the
+ * `rankCompare` preorder, where there may be internal runs tied for the same
+ * rank, into an iterable that resolves those ties using `fullCompare`.
+ *
+ * @template [V=unknown]
+ * @param {Iterable<[Key, V]>} rankSorted
+ * @param {RankCompare} rankCompare
+ * @param {FullCompare} fullCompare
+ * @returns {IterableIterator<[Key, V]>}
+ */
+const generateFullSortedEntries = (rankSorted, rankCompare, fullCompare) => {
+  /** @type {Array<[Key, V]>} */
+  let arrEntries;
+  if (Array.isArray(rankSorted)) {
+    arrEntries = rankSorted;
+  } else {
+    arrEntries = harden([...rankSorted]);
+  }
```

Comment:
Huh. Or rather, the other way around since this one immediately converts to array anyway?

Reply:
The motivation for `generateFullSortedEntries` accepting an iterable is direct use of `getCopyMapEntries`, although it can easily be refactored to instead accept an array if we reify a function for producing that kind of array from CopyMap input and give it a slightly worse name (e.g., `getCopyMapEntryArray = m => { assertCopyMap(m); const { payload: { keys, values } } = m; return keys.map((key, i) => [key, values[i]]); }`), which would eliminate the unnecessary copying of CopySet and CopyBag contents.

Reply:
I'd like to see that refactor but I'm fine either way for now. Resolving.

Reply:
I just went ahead with it: https://github.com/endojs/endo/pull/1737/commits/8f731ace5b1ac90fa74f7ba792309b2aa6809ba0